src/lib.rs

```diff
@@ -428,32 +260,20 @@ where
             });
         };
 
-        let worker = self.workers.get(&completion.task_queue).await;
-        let worker = if let Some(WorkerStatus::Live(w)) = worker.as_deref() {
-            w
-        } else {
-            return Err(CompleteActivityError::NoWorkerForQueue(
-                completion.task_queue,
-            ));
-        };
-
-        worker
-            .complete_activity(task_token, status, self.server_gateway.as_ref())
-            .await
+        let worker = self.worker(&completion.task_queue)?;
+        worker.complete_activity(task_token, status).await
     }
 
     fn record_activity_heartbeat(&self, details: ActivityHeartbeat) {
-        let workers = self.workers.clone();
-        // TODO: Ugh. Can we avoid making this async in a better way?
-        let _ = tokio::task::spawn(async move {
-            if let Some(WorkerStatus::Live(w)) = workers.get(&details.task_queue).await.as_deref() {
-                w.record_heartbeat(details)
-            }
-        });
+        if let Ok(w) = self.worker(&details.task_queue) {
+            w.record_heartbeat(details);
+        }
     }
 
-    fn request_workflow_eviction(&self, run_id: &str) {
-        self.wft_manager.request_eviction(run_id);
+    fn request_workflow_eviction(&self, task_queue: &str, run_id: &str) {
```

Comment:
I'd be fine with that but we talked about this before and said lang doesn't want to have to deal with this error since there's really nothing it can do about it, which I think makes sense.